File: windows8/src/cordova-win8/js/capture.js
 /**
 * Encapsulates all audio capture operation configuration options.
 */
function CaptureAudioOptions() {
    // Upper limit of sound clips user can record. Value must be equal or greater than 1.
    this.limit = 1;
    // Maximum duration of a single sound clip in seconds.
    this.duration = 0;
    // The selected audio mode. Must match with one of the elements in supportedAudioModes array.
    this.mode = null;
};

 /**
 * The CaptureError interface encapsulates all errors in the Capture API.
 */
function CaptureError(c) {
    this.code = c || null;
};

// Camera or microphone failed to capture image or sound.
CaptureError.CAPTURE_INTERNAL_ERR = 0;
// Camera application or audio capture application is currently serving other capture request.
CaptureError.CAPTURE_APPLICATION_BUSY = 1;
// Invalid use of the API (e.g. limit parameter has value less than one).
CaptureError.CAPTURE_INVALID_ARGUMENT = 2;
// User exited camera application or audio capture application before capturing anything.
CaptureError.CAPTURE_NO_MEDIA_FILES = 3;
// The requested capture operation is not supported.
CaptureError.CAPTURE_NOT_SUPPORTED = 20;


 /**
 * Encapsulates all image capture operation configuration options.
 */
function CaptureImageOptions() {
    // Upper limit of images user can take. Value must be equal or greater than 1.
    this.limit = 1;
    // The selected image mode. Must match with one of the elements in supportedImageModes array.
    this.mode = null;
};


 /**
 * Encapsulates all video capture operation configuration options.
 */
function CaptureVideoOptions() {
    // Upper limit of videos user can record. Value must be equal or greater than 1.
    this.limit = 1;
    // Maximum duration of a single video clip in seconds.
    this.duration = 0;
    // The selected video mode. Must match with one of the elements in supportedVideoModes array.
    this.mode = null;
};

/**
 * Represents a single file.
 *
 * name {DOMString} name of the file, without path information
 * fullPath {DOMString} the full path of the file, including the name
 * type {DOMString} mime type
 * lastModifiedDate {Date} last modified date
 * size {Number} size of the file in bytes
 */
function MediaFile(name, fullPath, type, lastModifiedDate, size) {
    MediaFile.__super__.constructor.apply(this, arguments);
};

utils.extend(MediaFile, File);

/**
 * Request capture format data for a specific file and type
 *
 * @param {Function} successCB
 * @param {Function} errorCB
 */
MediaFile.prototype.getFormatData = function (successCallback, errorCallback) {
    var contentType = this.type;
    if (typeof this.fullPath === "undefined" || this.fullPath === null) {
        errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
    } else {
        Windows.Storage.StorageFile.getFileFromPathAsync(this.fullPath).then(function (storageFile) {
            var mediaTypeFlag = String(contentType).split("/")[0].toLowerCase();
            if (mediaTypeFlag === "audio") {
                storageFile.properties.getMusicPropertiesAsync().then(function (audioProperties) {
                    successCallback(new MediaFileData(null, audioProperties.bitrate, 0, 0, audioProperties.duration/1000));
                }, function () {
                    errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
                })
            }
            else if (mediaTypeFlag === "video") {
                storageFile.properties.getVideoPropertiesAsync().then(function (videoProperties) {
                    successCallback(new MediaFileData(null, videoProperties.bitrate, videoProperties.height, videoProperties.width, videoProperties.duration/1000));
                }, function () {
                    errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
                })
            }
            else if (mediaTypeFlag === "image") {
                storageFile.properties.getImagePropertiesAsync().then(function (imageProperties) {
                    successCallback(new MediaFileData(null, 0, imageProperties.height, imageProperties.width, 0));
                }, function () {
                    errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
                })
            }
            else { errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT)) }
        }, function () {
            errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
        })
    }
};

/**
 * MediaFileData encapsulates format information of a media file.
 *
 * @param {DOMString} codecs
 * @param {long} bitrate
 * @param {long} height
 * @param {long} width
 * @param {float} duration
 */
function MediaFileData (codecs, bitrate, height, width, duration) {
    this.codecs = codecs || null;
    this.bitrate = bitrate || 0;
    this.height = height || 0;
    this.width = width || 0;
    this.duration = duration || 0;
};
/**
 * The Capture interface exposes an interface to the camera and microphone of the hosting device.
 */
function Capture() {
    this.supportedAudioModes = [];
    this.supportedImageModes = ["image/jpeg", "image/png"];
    this.supportedVideoModes = ["video/mp4", "video/wmv"];
}

/**
 * Launch audio recorder application for recording audio clip(s).
 *
 * @param {Function} successCB
 * @param {Function} errorCB
 * @param {CaptureAudioOptions} options
 */

// No UI support. The duration of the audio recording.
var cameraCaptureAudioDuration;
Capture.prototype.captureAudio = function (successCallback, errorCallback, options) {
    var audioOptions = new CaptureAudioOptions();
    if (options.duration && options.duration > 0) {
        audioOptions.duration = options.duration;
        cameraCaptureAudioDuration = audioOptions.duration;
    } else {
        errorCallback(new CaptureError(CaptureError.CAPTURE_INVALID_ARGUMENT));
        return;
    }
    var mediaCaputreSettings;
    var initCaptureSettings = function () {
        mediaCaputreSettings = null;
        mediaCaputreSettings = new Windows.Media.Capture.MediaCaptureInitializationSettings();
        mediaCaputreSettings.streamingCaptureMode = Windows.Media.Capture.StreamingCaptureMode.audio;
    }
    initCaptureSettings();
    var mediaCapture = new Windows.Media.Capture.MediaCapture();
    mediaCapture.initializeAsync(mediaCaputreSettings).done(function () {
        Windows.Storage.KnownFolders.musicLibrary.createFileAsync("captureAudio.mp3", Windows.Storage.NameCollisionOption.generateUniqueName).then(function (storageFile) {
            var mediaEncodingProfile = new Windows.Media.MediaProperties.MediaEncodingProfile.createMp3(Windows.Media.MediaProperties.AudioEncodingQuality.auto);
            var stopRecord = function () {
                mediaCapture.stopRecordAsync().then(function (result) {
                    storageFile.getBasicPropertiesAsync().then(function (basicProperties) {
                        successCallback(new MediaFile(storageFile.name, storageFile.path, storageFile.contentType, basicProperties.dateModified, basicProperties.size));
                    }, function () {
                        errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES));
                    })
                }, function () { errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES)); })
            }
            mediaCapture.startRecordToStorageFileAsync(mediaEncodingProfile, storageFile).then(function () {
                setTimeout(stopRecord, cameraCaptureAudioDuration*1000);
            }, function () { errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES)); })
        }, function () { errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES)); })
    })
};

/**
 * Launch camera application for taking image(s).
 *
 * @param {Function} successCB
 * @param {Function} errorCB
 * @param {CaptureImageOptions} options
 */
Capture.prototype.captureImage = function (successCallback, errorCallback, options) {
    var imageOptions = new CaptureImageOptions();
    var cameraCaptureUI = new Windows.Media.Capture.CameraCaptureUI();
    cameraCaptureUI.photoSettings.allowCropping = true;
    cameraCaptureUI.photoSettings.maxResolution = Windows.Media.Capture.CameraCaptureUIMaxPhotoResolution.highestAvailable;
    cameraCaptureUI.photoSettings.format = Windows.Media.Capture.CameraCaptureUIPhotoFormat.jpeg;
    cameraCaptureUI.captureFileAsync(Windows.Media.Capture.CameraCaptureUIMode.photo).then(function (file) {
        file.moveAsync(Windows.Storage.KnownFolders.picturesLibrary, "cameraCaptureImage.jpg", Windows.Storage.NameCollisionOption.generateUniqueName).then(function () {
            file.getBasicPropertiesAsync().then(function (basicProperties) {
                successCallback(new MediaFile(file.name, file.path, file.contentType, basicProperties.dateModified, basicProperties.size));
            }, function () {
                errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES));
            })
        }, function () {
            errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES));
        });
    }, function () { errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES)); })
};

/**
 * Launch device camera application for recording video(s).
 *
 * @param {Function} successCB
 * @param {Function} errorCB
 * @param {CaptureVideoOptions} options
 */
Capture.prototype.captureVideo = function (successCallback, errorCallback, options) {
    var videoOptions = new CaptureVideoOptions();
    if (options.duration && options.duration > 0) {
        videoOptions.duration = options.duration;
    }
    if (options.limit > 1) {
        videoOptions.limit = options.limit;
    }
    var cameraCaptureUI = new Windows.Media.Capture.CameraCaptureUI();
    cameraCaptureUI.videoSettings.allowTrimming = true;
    cameraCaptureUI.videoSettings.format = Windows.Media.Capture.CameraCaptureUIVideoFormat.mp4;
    cameraCaptureUI.videoSettings.maxDurationInSeconds = videoOptions.duration;
    cameraCaptureUI.captureFileAsync(Windows.Media.Capture.CameraCaptureUIMode.video).then(function (file) {
        file.moveAsync(Windows.Storage.KnownFolders.videosLibrary, "cameraCaptureVedio.mp4", Windows.Storage.NameCollisionOption.generateUniqueName).then(function () {
            file.getBasicPropertiesAsync().then(function (basicProperties) {
                successCallback(new MediaFile(file.name, file.path, file.contentType, basicProperties.dateModified, basicProperties.size));
            }, function () {
                errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES));
            })
        }, function () {
            errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES));
        });
    }, function () { errorCallback(new CaptureError(CaptureError.CAPTURE_NO_MEDIA_FILES)); })
};


/**
 * Encapsulates a set of parameters that the capture device supports.
 */
function ConfigurationData() {
    // The ASCII-encoded string in lower case representing the media type.
    this.type = null;
    // The height attribute represents height of the image or video in pixels.
    // In the case of a sound clip this attribute has value 0.
    this.height = 0;
    // The width attribute represents width of the image or video in pixels.
    // In the case of a sound clip this attribute has value 0
    this.width = 0;
}

// Just for test(because of the unimplementation of channnel and device)
function Device() { }

if (typeof navigator.device === "undefined") {
    navigator.device = new Device();
}

if (typeof navigator.device.capture === "undefined") {
    navigator.device.capture = new Capture();
}

File: Windows8PhonegapWinRT/Windows8PhonegapWinRT/Commands/Capture.cs
/*  
	Licensed under the Apache License, Version 2.0 (the "License");
	you may not use this file except in compliance with the License.
	You may obtain a copy of the License at
	
	http://www.apache.org/licenses/LICENSE-2.0
	
	Unless required by applicable law or agreed to in writing, software
	distributed under the License is distributed on an "AS IS" BASIS,
	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
	See the License for the specific language governing permissions and
	limitations under the License.
*/


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.Serialization;
using Windows.UI.Xaml.Media.Imaging;
using Windows.Storage;
using Windows.Storage.FileProperties;
using Windows.Media.Capture;
using Windows.Storage.Streams;
using Windows.Media.MediaProperties;

namespace Windows8PhonegapWinRT.Commands
{
    public class Capture : BaseCommand
    {
        #region Internal classes (options and resultant objects)

        /// <summary>
        /// Represents captureImage action options.
        /// </summary>
        [DataContract]
        public class CaptureImageOptions
        {
            /// <summary>
            /// The maximum number of images the device user can capture in a single capture operation. The value must be greater than or equal to 1 (default to 1).
            /// </summary>
            [DataMember(IsRequired = false, Name = "limit")]
            public int Limit { get; set; }

            public static CaptureImageOptions Default
            {
                get { return new CaptureImageOptions() { Limit = 1 }; }
            }
        }

        /// <summary>
        /// Represents captureVideo action options.
        /// </summary>
        [DataContract]
        public class CaptureVideoOptions
        {
            /// <summary>
            /// The maximum number of video files the device user can capture in a single capture operation. The value must be greater than or equal to 1 (default to 1).
            /// </summary>
            [DataMember(IsRequired = false, Name = "limit")]
            public int Limit { get; set; }

            public static CaptureVideoOptions Default
            {
                get { return new CaptureVideoOptions() { Limit = 1 }; }
            }

        } 

        /// <summary>
        /// Represents captureAudio action options.
        /// </summary>
        [DataContract]
        public class CaptureAudioOptions
        {
            /// <summary>
            /// The maximum number of audio files the device user can capture in a single capture operation. The value must be greater than or equal to 1 (defaults to 1).
            /// </summary>
            [DataMember(IsRequired = false, Name = "limit")]
            public int Limit { get; set; }

            public static CaptureAudioOptions Default
            {
                get { return new CaptureAudioOptions() { Limit = 1 }; }
            }
        }


        /// <summary>
        /// Store image info
        /// </summary>
        [DataContract]
        public class MediaFile
        {
            [DataMember(Name = "name")]
            public string FileName { get; set; }

            [DataMember(Name = "fullPath")]
            public string FilePath { get; set; }

            [DataMember(Name = "type")]
            public string Type { get; set; }

            [DataMember(Name = "lastModifiedDate")]
            public string LastModifiedDate { get; set; }

            [DataMember(Name = "size")]
            public long Size { get; set; }

            public MediaFile(string filepath, string imgType, string name, long size, string lastModDate)
            {
                //var basicProperties = imgFile.GetBasicPropertiesAsync();

                this.FilePath = filepath;
                this.FileName = name;
                this.Type = imgType;
                this.Size = size;
                this.LastModifiedDate = lastModDate;

            }
        }
        #endregion

        /////<summary>
        /////Folder to store captured images
        /////</summary>
        //private string isoFolder = "CapturedImagesCache";

        ///<summary>
        ///Capture Image options
        ///</summary>
        private CaptureImageOptions captureImageOptions;

        ///<summary>
        ///Capture Video options
        ///</summary>
        protected CaptureVideoOptions captureVideoOptions;

        ///<summary>
        ///Capture Video options
        ///</summary>
        protected CaptureAudioOptions captureAudioOptions;

        ///<summary>
        /// Use to open camera application
        ///</summary>
        private CameraCaptureUI cameraTask;

        /// <summary>
        /// Used for audio recording
        /// </summary>
        private MediaCapture audioCaptureTask;

        /// <summary>
        /// Time in second for audio recording
        /// </summary>
        double captureAudioDuration = 5;

        ///<summary>
        ///Stores information about captured files
        ///</summary>
        List<MediaFile> files = new List<MediaFile>();


        ///<summary>
        ///Lunch default camera application to capture image
        ///</summary>
        ///<param name="options">may contains limit of mode parameters</param>
        ///
        public async void captureImage(string options)
        {
            try
            {
                try
                {
                    string args = JSON.JsonHelper.Deserialize<string[]>(options)[0];
                    this.captureImageOptions = String.IsNullOrEmpty(args) ? CaptureImageOptions.Default : JSON.JsonHelper.Deserialize<CaptureImageOptions>(args);
                }
                catch (Exception ex)
                {
                    this.DispatchCommandResult(new PluginResult(PluginResult.Status.JSON_EXCEPTION, ex.Message));
                    return;
                }

                cameraTask = new CameraCaptureUI();
                cameraTask.PhotoSettings.AllowCropping = true;
                cameraTask.PhotoSettings.MaxResolution = CameraCaptureUIMaxPhotoResolution.HighestAvailable;
                cameraTask.PhotoSettings.Format = CameraCaptureUIPhotoFormat.Jpeg;

                StorageFile picture = await cameraTask.CaptureFileAsync(CameraCaptureUIMode.Photo);

                if (picture != null)
                {
                    await picture.CopyAsync(Windows.Storage.KnownFolders.PicturesLibrary, picture.Name, Windows.Storage.NameCollisionOption.GenerateUniqueName);

                    long size = 0;
                    string modifiedDate = "";
                    var tasks = new List<Task<BasicProperties>>();
                    tasks.Add(picture.GetBasicPropertiesAsync().AsTask());
                    var result = await Task.WhenAll(tasks);
                    foreach (var prop in result)
                    {
                        size = (long)prop.Size;
                        modifiedDate = prop.DateModified.ToString();
                    }

                    string imagePathOrContent = string.Empty;
                    var readStream = await picture.OpenAsync(FileAccessMode.Read);
                    var inputStream = readStream.GetInputStreamAt(0);
                    var dataReaderFile = new DataReader(inputStream);
                    var numByteLoaded = await dataReaderFile.LoadAsync((uint)readStream.Size);
                    var byteString = new byte[numByteLoaded];
                    var imageBase64String = "";
                    dataReaderFile.ReadBytes(byteString);
                    imageBase64String = Convert.ToBase64String(byteString);
                    imagePathOrContent = "data:image/jpeg;base64," + imageBase64String;

                    MediaFile data = new MediaFile(imagePathOrContent, picture.ContentType, picture.Name, size, modifiedDate);

                    this.files.Add(data);

                    if (files.Count < this.captureImageOptions.Limit)
                    {
                        //dosomething here
                    }
                    else
                    {
                        DispatchCommandResult(new PluginResult(PluginResult.Status.OK, files));
                        files.Clear();
                    }

                }
                else
                {
                    DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, "Error in capturing image"));
                }
            }
            catch (Exception ex)
            {
                var error = ex.Message.ToString();
                DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, "Error capturing image."));
            }

        }

        /// <summary>
        /// Launches our own video recording control to capture video
        /// </summary>
        /// <param name="options">may contains additional parameters</param>
        public async void captureVideo(string options)
        {
            try
            {
                try
                {
                    string args = JSON.JsonHelper.Deserialize<string[]>(options)[0];
                    this.captureVideoOptions = String.IsNullOrEmpty(args) ? CaptureVideoOptions.Default : JSON.JsonHelper.Deserialize<CaptureVideoOptions>(args);
                }
                catch (Exception ex)
                {
                    this.DispatchCommandResult(new PluginResult(PluginResult.Status.JSON_EXCEPTION, ex.Message));
                    return;
                }

                cameraTask = new CameraCaptureUI();
                cameraTask.VideoSettings.AllowTrimming = true;
                cameraTask.VideoSettings.Format = CameraCaptureUIVideoFormat.Mp4;

                StorageFile video = await cameraTask.CaptureFileAsync(CameraCaptureUIMode.Video);

                if (video != null)
                {
                    await video.CopyAsync(Windows.Storage.KnownFolders.VideosLibrary, video.Name, Windows.Storage.NameCollisionOption.GenerateUniqueName);

                    long size = 0;
                    string modifiedDate = "";
                    var tasks = new List<Task<BasicProperties>>();
                    tasks.Add(video.GetBasicPropertiesAsync().AsTask());
                    var result = await Task.WhenAll(tasks);
                    foreach (var prop in result)
                    {
                        size = (long)prop.Size;
                        modifiedDate = prop.DateModified.ToString();
                    }

                    MediaFile data = new MediaFile(video.Path, video.ContentType, video.Name, size, modifiedDate);

                    this.files.Add(data);

                    if (files.Count < this.captureVideoOptions.Limit)
                    {
                        //dosomething here
                    }
                    else
                    {
                        DispatchCommandResult(new PluginResult(PluginResult.Status.OK, files));
                        files.Clear();
                    }
                }
                else
                {
                    DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, "Error in capturing video"));
                }

            }
            catch (Exception ex)
            {
                DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, ex.Message));
            }
        }

        public async void captureAudio(string options)
        {
            try
            {
                try
                {
                    string args = JSON.JsonHelper.Deserialize<string[]>(options)[0];
                    this.captureAudioOptions = String.IsNullOrEmpty(args) ? CaptureAudioOptions.Default : JSON.JsonHelper.Deserialize<CaptureAudioOptions>(args);
                }
                catch (Exception ex)
                {
                    this.DispatchCommandResult(new PluginResult(PluginResult.Status.JSON_EXCEPTION, ex.Message));
                    return;
                }

                var mediaCaputreSettings = new MediaCaptureInitializationSettings();
                mediaCaputreSettings.StreamingCaptureMode = StreamingCaptureMode.Audio;
                audioCaptureTask = new MediaCapture();
                await audioCaptureTask.InitializeAsync(mediaCaputreSettings);

                var mediaEncodingProfile = MediaEncodingProfile.CreateMp3(AudioEncodingQuality.Auto);
                var storageFile = await KnownFolders.MusicLibrary.CreateFileAsync("captureAudio.mp3", CreationCollisionOption.GenerateUniqueName);
                
                var timer = new Windows.UI.Xaml.DispatcherTimer();
                timer.Tick += async delegate(object sender, object e)
                {
                    timer.Stop();
                    await audioCaptureTask.StopRecordAsync();

                    if (storageFile != null)
                    {
                        long size = 0;
                        string modifiedDate = "";
                        var tasks = new List<Task<BasicProperties>>();
                        tasks.Add(storageFile.GetBasicPropertiesAsync().AsTask());
                        var result = await Task.WhenAll(tasks);
                        foreach (var prop in result)
                        {
                            size = (long)prop.Size;
                            modifiedDate = prop.DateModified.ToString();
                        }

                        string imagePathOrContent = string.Empty;
                        var readStream = await storageFile.OpenAsync(FileAccessMode.Read);
                        var inputStream = readStream.GetInputStreamAt(0);
                        var dataReaderFile = new DataReader(inputStream);
                        var numByteLoaded = await dataReaderFile.LoadAsync((uint)readStream.Size);
                        var byteString = new byte[numByteLoaded];
                        var imageBase64String = "";
                        dataReaderFile.ReadBytes(byteString);
                        imageBase64String = Convert.ToBase64String(byteString);
                        imagePathOrContent = "data:audio/mpeg;base64," + imageBase64String;

                        MediaFile data = new MediaFile(imagePathOrContent, storageFile.ContentType, storageFile.Name, size, modifiedDate);

                        this.files.Add(data);

                        if (files.Count < this.captureAudioOptions.Limit)
                        {
                            //dosomething here
                        }
                        else
                        {
                            DispatchCommandResult(new PluginResult(PluginResult.Status.OK, files));
                            files.Clear();
                        }
                    }
                };
                timer.Interval = TimeSpan.FromMilliseconds(captureAudioDuration * 1000);
                

                await audioCaptureTask.StartRecordToStorageFileAsync(mediaEncodingProfile, storageFile);
                timer.Start();

                   
                
                
            }
            catch (Exception ex)
            {
                DispatchCommandResult(new PluginResult(PluginResult.Status.ERROR, ex.Message));
            }
        }

    }
}
